plugins/test/dynamic_test.cc

```diff
@@ -392,7 +402,66 @@ void DynamicTest::CheckSideEffects(const std::string& phase,
 void DynamicTest::CheckPhaseResults(const std::string& phase,
                                     const pb::Expectation& expect,
                                     const TestContext& context,
-                                    const TestHttpContext::Result& result) {
+                                    const TestHttpContext::Result& result,
+                                    bool* immediate_response_sent) {
+  if (*immediate_response_sent &&
```

Comment:
IMO tracking this boolean is unnecessarily complex. If you needed it, you could instead expose the new immediate_response_ boolean tracked inside TestHttpContext.

But my preferred approach is to not change anything in DynamicTest: the changes you made in TestHttpContext are sufficient to make any post-immediate-response tests fail, because we never invoke those callbacks anymore.